sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -23,41 +29,64 @@ inline void SetLibcurlOption(
         errorMessage + ". " + std::string(curl_easy_strerror(result)));
   }
 }
-} // namespace
-
-using namespace Azure::Core::Http;
 
-// To wait for a socket to be ready to be read/write
-// Method From: https://github.com/curl/curl/blob/master/docs/examples/sendrecv.c#L32
-// Copyright (c) 1996 - 2020, Daniel Stenberg, <daniel@haxx.se>
-static int WaitForSocketReady(curl_socket_t sockfd, int for_recv, long timeout_ms)
+enum class PollSocketDirection
+{
+  Read = 1,
+  Write = 2,
+};
+
+/**
+ * @brief Use poll from OS to check if socket is ready to be read or written.
+ *
+ * @param socketFileDescriptor socket descriptor.
+ * @param direction poll events for read or write socket.
+ * @param timeout  return if polling for more than \p timeout
+ * @return int with negative 1 upon any error, 0 on timeout or greater than zero if events were
+ * detected (socket ready to be written/read)
+ */
+int pollSocketUntilEventOrTimeout(
+    curl_socket_t socketFileDescriptor,
+    PollSocketDirection direction,
+    long timeout)
 {
-  struct timeval tv;
-  fd_set infd, outfd, errfd;
-  int res;
-
-  tv.tv_sec = timeout_ms / 1000;
-  tv.tv_usec = (timeout_ms % 1000) * 1000;
 
-  FD_ZERO(&infd);
-  FD_ZERO(&outfd);
-  FD_ZERO(&errfd);
+#ifndef POSIX
+#ifndef WINDOWS
+  // platform does not support Poll().
+  // TODO. Legacy select() for other platforms?
+  throw Azure::Core::Http::TransportException(
+      "Error while sending request. Platform does not support Poll()");
+#endif
+#endif
 
-  FD_SET(sockfd, &errfd); /* always check for error */
+  struct pollfd poller;
+  poller.fd = socketFileDescriptor;
 
-  if (for_recv)
+  // set direction
+  if (direction == PollSocketDirection::Read)
   {
-    FD_SET(sockfd, &infd);
+    poller.events = POLLIN;
   }
   else
   {
-    FD_SET(sockfd, &outfd);
+    poller.events = POLLOUT;
```

Comment:
How to handle error? do we need to poll on POLLERR events?

Reply:
Error is handled from whoever calls this function.
If this function `pollSocketUntilEventOrTimeout` returns -1 it means error, 
if it returns 0 it means timeout,
greater than 0 means events found (socket ready to read or write)